### PR TITLE
Change display-mode without restart

### PR DESCRIPTION
**Pull request by huwpascoe**
Working prototype of display mode change without restarting. (Live display adjustments https://github.com/OpenRA/OpenRA/issues/3774) Type the new res in the display settings and go back/to another settings tab.

Need help to finish:
Not all ingame widgets are initialized properly
Nitpicking the display settings UI - I think a dropdown resolution changer would be better
Probably don't need help with but haven't included yet: fullscreen switching, hardware cursor switching

### CODE REVIEW

**Comment by pchote** [comment]
Unfortunately changing the resolution is the easy part.  95% of the work in this feature is fixing the widget code to support relayout.  I suggest doing this by having code to adjust the existing widget bounds.  There is too much internal state for throwing out and rebuilding the widget tree to work unless you wanted to limit this feature to only work from the settings menu on the shellmap, which defeats the purpose.

**Comment by penev92** [comment]
> Nitpicking the display settings UI - I think a dropdown resolution changer would be better

Please no. I run it in windowed mode at 1024x736, which is something you won't find in any dropdown in any game.

File: OpenRA.Game/Game.cs [hunk 1]
```diff
@@ -69,6 +69,15 @@ static void JoinInner(OrderManager om)
 			ConnectionStateChanged(OrderManager);
 		}
 
+		public static void ChangeRes(Size size)
```
Comment (pchote) [inline OpenRA.Game/Game.cs]: This isn't going to work in the long term: you can't know ahead of time everything that needs to respond to resolution changes.

You should instead set up an `OnResolutionChange` event in the Renderer class that everything else attaches callbacks to.

File: OpenRA.Game/Graphics/IGraphicsDevice.cs [hunk 2]
```diff
@@ -55,6 +55,7 @@ public interface IGraphicsDevice : IDisposable
 		IShader CreateShader(string name);
 
 		Size WindowSize { get; }
+		void ChangeRes(Size size, WindowMode windowMode);
```
Comment (pchote) [inline OpenRA.Game/Graphics/IGraphicsDevice.cs]: `SetWindowSize(Size size)` please.  This doesn't use the `windowMode` arg, and there's no need to shorten "resolution" to "res" everywhere

File: OpenRA.Game/Graphics/Viewport.cs [hunk 3]
```diff
@@ -214,7 +214,14 @@ public void Scroll(float2 delta, bool ignoreBorders)
 		}
 
 		// Rectangle (in viewport coords) that contains things to be drawn
-		static readonly Rectangle ScreenClip = Rectangle.FromLTRB(0, 0, Game.Renderer.Resolution.Width, Game.Renderer.Resolution.Height);
+		static Rectangle ScreenClip
+		{
+			get
+			{
+				return Rectangle.FromLTRB(0, 0, Game.Renderer.Resolution.Width, Game.Renderer.Resolution.Height);
```
Comment (pchote) [inline OpenRA.Game/Graphics/Viewport.cs]: This will create additional GC pressure by creating and discarding many more objects.  You can keep this cached, and then use the OnResolutionChange event to update its value.

**Comment by huwpascoe** [comment]
@penev92 Please no. I run it in windowed mode at 1024x736, which is something you won't find in any dropdown in any game. Well lets have both a dropdown and a manual input : )

**Comment by x-a-n-a-x** [comment]
That wouldn't shock me: https://cloud.githubusercontent.com/assets/1732773/10564093/4c0f86c0-75a6-11e5-8773-ebd4104ecc10.png

**Comment by huwpascoe** [comment]
There, got all the display modes working too!
openrares

The UI seems to be able to adjust to the changes now thanks to OnChangeResolution() although there are a few more bugs to find I'm sure.

I don't really understand what the hardware cursor is doing though if anyone has an explanation/ideas?

File: OpenRA.Game/Renderer.cs [hunk 4]
```diff
@@ -91,6 +85,24 @@ static IGraphicsDevice CreateDevice(Assembly platformDll, int width, int height,
 			throw new InvalidOperationException("Renderer DLL is missing RendererAttribute to tell us what type to use!");
 		}
 
+		public readonly Size MinimumResolution = new Size(800, 600);
+
+		public readonly Size[] SuggestedResolutions = new Size[]
```
Comment (pchote) [inline OpenRA.Game/Renderer.cs]: These are all common screen resolutions, so it doesn't make sense to offer them.  There is no room for the window decorations!

File: OpenRA.Game/Graphics/IGraphicsDevice.cs [hunk 5]
```diff
@@ -135,6 +137,6 @@ public enum WindowMode
 	{
 		Windowed,
 		Fullscreen,
-		PseudoFullscreen,
+		NativeFullscreen,
```
Comment (pchote) [inline OpenRA.Game/Graphics/IGraphicsDevice.cs]: You've introduced a terminology change here.  This should be at least in its own commit, but preferably in its own pr.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs [hunk 6]
```diff
@@ -182,13 +173,25 @@ Action InitDisplayPanel(Widget panel)
 			var cursorDoubleIsChecked = cursorDoubleCheckbox.IsChecked;
 			cursorDoubleCheckbox.IsChecked = () => !cursorDoubleCheckbox.IsDisabled() && cursorDoubleIsChecked();
 
-			panel.Get("WINDOW_RESOLUTION").IsVisible = () => ds.Mode == WindowMode.Windowed;
+			panel.Get("WINDOW_RESOLUTION").IsVisible = () => ds.Mode != WindowMode.NativeFullscreen;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: Erm, what's going on here?  The resolution control only makes sense for the windowed mode.

File: OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs [hunk 7]
```diff
@@ -98,6 +86,47 @@ public Sdl2GraphicsDevice(Size windowSize, WindowMode windowMode)
 
 			SDL.SDL_SetModState(SDL.SDL_Keymod.KMOD_NONE);
 			input = new Sdl2Input();
+
+			Clear();
+			Present();
+			SetWindowSize(WindowSize, windowMode);
+		}
+
+		public void SetWindowSize(Size windowSize, WindowMode windowMode)
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: You're now handling the window mode as well, so calling this `SetWindowSize` doesn't make sense.  It should be changed to `SetWindowMode`.

File: OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs [hunk 8]
```diff
@@ -98,6 +86,47 @@ public Sdl2GraphicsDevice(Size windowSize, WindowMode windowMode)
...
+				SDL.SDL_SetWindowFullscreen(window, (uint)SDL.SDL_WindowFlags.SDL_WINDOW_FULLSCREEN);
+			}
+
+			if (windowMode == WindowMode.NativeFullscreen)
+				SDL.SDL_SetWindowFullscreen(window, (uint)SDL.SDL_WindowFlags.SDL_WINDOW_FULLSCREEN_DESKTOP);
+
+			if (windowMode == WindowMode.Windowed)
+				SDL.SDL_SetWindowPosition(window, SDL.SDL_WINDOWPOS_CENTERED, SDL.SDL_WINDOWPOS_CENTERED);
+			else
+				SDL.SDL_SetWindowPosition(window, 0, 0);
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: Be careful about / please carry over the commented workaround for OSX from the code you've removed.

**Comment by pchote** [comment]
This is already starting to turn into a monster pr with multiple distinct changes.  Please keep the scope small, and split out smaller prs as needed.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs (hunk 6 above)
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: What? Definitely no, fullscreen (or pseudo-notsure of a sudden which I used) resolution change can be done via editing the settings.yaml on bleed. It makes sense as a form of full zooming in.

**Comment by pchote** [comment]
> The UI seems to be able to adjust to the changes now thanks to OnChangeResolution() although there are a few more bugs to find I'm sure.

The loadscreen geometry is wrong, and the ingame UI breaks if you change resolution from inside a game.

File: OpenRA.Game/Widgets/Widget.cs [hunk 9]
```diff
@@ -237,6 +241,15 @@ public virtual void Initialize(WidgetArgs args)
 								   height);
 		}
 
+		protected void OnResolutionChange()
+		{
+			if (widgetArgs != null)
+				Initialize(widgetArgs);
```
Comment (pchote) [inline OpenRA.Game/Widgets/Widget.cs]: Widgets do more than just lay out their geometry in Inititialize.  Calling this multiple times is a bad idea.  You should instead implement a new `OnResolutionChangeInner` method that each widget can implement only the things it needs on resize (most widgets will be able to use a default implementation on the base widget).

**Comment by pchote** [comment]
Its going to take a bit more work to get this mergable, but its a promising start!

**Comment by pchote** [comment]
Changing to a custom resolution in Fullscreen (not native) under OSX gives bogus geometry, chopping off part of the screen and giving different coordinates for rendering and inputs (so there is a constant offset between where the cursor is rendered and where click events are fired).  This is probably an SDL bug rather than something you did.

File: OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs (hunk 6 above)
Comment (huwpascoe) [inline OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs]: True full screen mode allows resolution change, this is a highly desired feature!

File: OpenRA.Game/Renderer.cs (hunk 4 above)
Comment (huwpascoe) [inline OpenRA.Game/Renderer.cs]: That gives me an idea actually. I'll get back to that one

File: OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs (hunk 8 above)
Comment (huwpascoe) [inline OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: This is no longer a workaround but required by all systems. Should I still grab the comment?

**Comment by huwpascoe** [comment]
Can you provide a screenshot of the osx bug?

File: OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs (hunk 8 above)
Comment (huwpascoe) [inline OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: Er required for correct mode changing I mean

**Comment by pchote** [comment]
screen shot 2015-10-18 at 23 05 36

Note that the (hardware rendered) cursor was positioned near the "OpenRA" text, while the game thought it was over the Multiplayer button.  This vertical offset matches the size of the black bar at the top of the screen.

**Comment by huwpascoe** [comment]
Closing this for now, it relies on another dependency which I'll probably make another PR for.

**Comment by pchote** [comment]
Adding to the future milestone so that we can find this again if you don't have time to reopen it.
